client/thin-replica-client/src/grpc_connection.cpp

```diff
@@ -107,16 +112,27 @@ void GrpcConnection::disconnect() {
   channel_.reset();
 }
 
+void GrpcConnection::checkAndReConnect(std::unique_ptr<GrpcConnectionConfig>& config) {
+  WriteLock write_lock(channel_protector_);
+  if (!isConnectedNoLock()) {
+    disconnect();
+    connect(config);
+  }
+}
+
 GrpcConnection::Result GrpcConnection::openDataStream(const SubscriptionRequest& request) {
+  ReadLock read_lock(channel_protector_);
```

Comment:
Could you please explain the rationale for using a `ReadLock` here?

Reply:
Thanks, @dartdart26 for the comments. In GrpcConnection the following 2 kinds of members are related and are protected by the channel_mutex_:
1) Channel -- This member is used by connect, reconnect, disconnect and isconnected family of functions.
2) trs_stub and rss_tub --> These stubs are created on top of the channel. They share the channel, so if we want to change the channel then we should make sure that stubs are not accessing the channel. The GRPC is implemented in such a manner that stubs can share a single channel.

Now all the functions which modify the channel directly will have to take a write lock. But when the stubs are used, we should make sure that no one modifies the channel. So we take read lock whenever we use the stub and write lock whenever we update channel. 

We should also note that we need to take write lock when we create the stub.

In all the open family of functions, we use the stub to call the API. So we take a read lock and we assume that channel_ is not nullptr.